# ==============================================================================
# REGRESSOR COMMIT DIFF
# ==============================================================================
# File: dom/indexedDB/IndexedDatabaseInlines.h
# Commit: 43dbad71664b
# Full Hash: 43dbad71664bb2c5779b54decd5133db076ec6a0
# Author: Simon Giesecke <sgiesecke@mozilla.com>
# Date: 2019-11-08 21:36:09
# Regressor Bug: 1168606
# File Overlap Count: 1
# Description:
#   Bug 1168606 - Replace pseudo-move constructor of StructuredCloneReadInfo by explicit DeserializeStructuredCloneReadInfo function. r=ttung,asuth
#   
#   Differential Revision: https://phabricator.services.mozilla.com/D46593
# ==============================================================================

diff -r 19188ca45113 -r 43dbad71664b dom/indexedDB/IndexedDatabaseInlines.h
--- a/dom/indexedDB/IndexedDatabaseInlines.h	Fri Nov 08 14:28:39 2019 +0000
+++ b/dom/indexedDB/IndexedDatabaseInlines.h	Fri Nov 08 13:22:08 2019 +0000
@@ -21,7 +21,15 @@
 namespace dom {
 namespace indexedDB {
 
-inline StructuredCloneFile::StructuredCloneFile() : mType(eBlob) {
+inline StructuredCloneFile::StructuredCloneFile(FileType aType,
+                                                RefPtr<Blob> aBlob)
+    : mBlob{std::move(aBlob)}, mType{aType} {
+  MOZ_COUNT_CTOR(StructuredCloneFile);
+}
+
+inline StructuredCloneFile::StructuredCloneFile(
+    RefPtr<IDBMutableFile> aMutableFile)
+    : mMutableFile{std::move(aMutableFile)}, mType{eMutableFile} {
   MOZ_COUNT_CTOR(StructuredCloneFile);
 }
 
@@ -47,25 +55,28 @@
           JS::StructuredCloneScope::DifferentProcessForIndexedDB) {}
 
 inline StructuredCloneReadInfo::StructuredCloneReadInfo(
-    StructuredCloneReadInfo&& aCloneReadInfo)
-    : mData(std::move(aCloneReadInfo.mData)) {
-  MOZ_ASSERT(&aCloneReadInfo != this);
+    JSStructuredCloneData&& aData, nsTArray<StructuredCloneFile> aFiles,
+    IDBDatabase* aDatabase, bool aHasPreprocessInfo)
+    : mData{std::move(aData)},
+      mFiles{std::move(aFiles)},
+      mDatabase{aDatabase},
+      mHasPreprocessInfo{aHasPreprocessInfo} {
+  MOZ_COUNT_CTOR(StructuredCloneReadInfo);
+}
+
+#ifdef NS_BUILD_REFCNT_LOGGING
+inline StructuredCloneReadInfo::StructuredCloneReadInfo(
+    StructuredCloneReadInfo&& aOther) noexcept
+    : mData(std::move(aOther.mData)) {
+  MOZ_ASSERT(&aOther != this);
   MOZ_COUNT_CTOR(StructuredCloneReadInfo);
 
   mFiles.Clear();
-  mFiles.SwapElements(aCloneReadInfo.mFiles);
-  mDatabase = aCloneReadInfo.mDatabase;
-  aCloneReadInfo.mDatabase = nullptr;
-  mHasPreprocessInfo = aCloneReadInfo.mHasPreprocessInfo;
-  aCloneReadInfo.mHasPreprocessInfo = false;
-}
-
-inline StructuredCloneReadInfo::StructuredCloneReadInfo(
-    SerializedStructuredCloneReadInfo&& aCloneReadInfo)
-    : mData(std::move(aCloneReadInfo.data().data)),
-      mDatabase(nullptr),
-      mHasPreprocessInfo(aCloneReadInfo.hasPreprocessInfo()) {
-  MOZ_COUNT_CTOR(StructuredCloneReadInfo);
+  mFiles.SwapElements(aOther.mFiles);
+  mDatabase = aOther.mDatabase;
+  aOther.mDatabase = nullptr;
+  mHasPreprocessInfo = aOther.mHasPreprocessInfo;
+  aOther.mHasPreprocessInfo = false;
 }
 
 inline StructuredCloneReadInfo::~StructuredCloneReadInfo() {
@@ -73,18 +84,19 @@
 }
 
 inline StructuredCloneReadInfo& StructuredCloneReadInfo::operator=(
-    StructuredCloneReadInfo&& aCloneReadInfo) {
-  MOZ_ASSERT(&aCloneReadInfo != this);
+    StructuredCloneReadInfo&& aOther) noexcept {
+  MOZ_ASSERT(&aOther != this);
 
-  mData = std::move(aCloneReadInfo.mData);
+  mData = std::move(aOther.mData);
   mFiles.Clear();
-  mFiles.SwapElements(aCloneReadInfo.mFiles);
-  mDatabase = aCloneReadInfo.mDatabase;
-  aCloneReadInfo.mDatabase = nullptr;
-  mHasPreprocessInfo = aCloneReadInfo.mHasPreprocessInfo;
-  aCloneReadInfo.mHasPreprocessInfo = false;
+  mFiles.SwapElements(aOther.mFiles);
+  mDatabase = aOther.mDatabase;
+  aOther.mDatabase = nullptr;
+  mHasPreprocessInfo = aOther.mHasPreprocessInfo;
+  aOther.mHasPreprocessInfo = false;
   return *this;
 }
+#endif
 
 inline size_t StructuredCloneReadInfo::Size() const {
   size_t size = mData.Size();
